ath79: add support for ZyXEL NWA1100-NH / NWA1121-NI / NWA1123-NI / NWA1123-AC

s-2 opened the pull request:
The NWA112x series are wall/ceiling-mounted PoE Business Access Points
in a compact, "smoke detector" style case.

Specifications:
 * AR9342, 16 MiB Flash, 64 MiB RAM, 802.11n 2T2R, 2.4 GHz
 * 1x Gigabit Ethernet, 802.3af PoE
 * NWA1123-NI has an additional AR9382 PCIe card, 802.11n 2T2R, 5 GHz
 * NWA1123-AC has an additional QCA9882 PCIe card, 802.11ac 2T2R, 5 GHz

Newer devices from the series (e.g. `NWA1123-AC v2` / `HD` / `Pro` etc.) use a different mtd layout and
bootloader (just as NWA512x, which even use the same hardware as NWA112x) and are not supported.

Installation:
* OEM Web UI is at 192.168.1.2
  login as `admin` with password `1234`
* Flash factory-XXX.bin

A model string `XXXX` needs to be present within the file name of the uploaded
image to be accepted by the OEM Web-based updater, the factory image is
named accordingly to save the user from the hassle of manual renaming.

TFTP Recovery:
* Open the case, connect to TTL UART port (this is the official method
  described by Zyxel, the reset button is useless during power-on)
* Extract factory image (.tar.bz2), serve `vmlinux_mi124_f1e.lzma.uImage`
  and `mi124_f1e-jffs2` via tftp at 192.168.1.10
* Interrupt uboot countdown, execute commands
  `run lk`
  `run lf`
  to flash the kernel / filesystem accordingly

Signed-off-by: Sebastian Schaper <openwrt@sebastianschaper.net>

s-2 committed:
ath79: add support for ZyXEL NWA1121-NI

Specifications:
 * AR9342, 16 MiB Flash, 64 MiB RAM, 802.11n 2T2R, 2.4 GHz
 * 1x Gigabit Ethernet (AR8035), 802.3af PoE

Installation:
* OEM Web UI is at 192.168.1.2
  login as `admin` with password `1234`
* Flash factory-AABJ.bin

The string `AABJ` needs to be present within the file name of the uploaded
image to be accepted by the OEM Web-based updater, the factory image is
named accordingly to save the user from the hassle of manual renaming.

TFTP Recovery:
* Open the case, connect to TTL UART port (this is the official method
  described by Zyxel, the reset button is useless during power-on)
* Extract factory image (.tar.bz2), serve `vmlinux_mi124_f1e.lzma.uImage`
  and `mi124_f1e-jffs2` via tftp at 192.168.1.10
* Interrupt uboot countdown, execute commands
  `run lk`
  `run lf`
  to flash the kernel / filesystem accordingly

MAC addresses as verified by OEM firmware:
use   address   source
LAN   *:cc      mib0 0x30 ('eth0mac'), art 0x1002 (label)
2g    *:cd      mib0 0x4b ('wifi0mac')

Signed-off-by: Sebastian Schaper <openwrt@sebastianschaper.net>

s-2 committed:
ath79: add support for ZyXEL NWA1123-NI

Specifications:
 * AR9342, 16 MiB Flash, 64 MiB RAM, 802.11n 2T2R, 2.4 GHz
 * AR9382 PCIe card, 802.11n 2T2R, 5 GHz
 * 1x Gigabit Ethernet (AR8035), 802.3af PoE

Installation:
* OEM Web UI is at 192.168.1.2
  login as `admin` with password `1234`
* Flash factory-AAEO.bin

The string `AAEO` needs to be present within the file name of the uploaded
image to be accepted by the OEM Web-based updater, the factory image is
named accordingly to save the user from the hassle of manual renaming.

TFTP Recovery:
* Open the case, connect to TTL UART port (this is the official method
  described by Zyxel, the reset button is useless during power-on)
* Extract factory image (.tar.bz2), serve `vmlinux_mi124_f1e.lzma.uImage`
  and `mi124_f1e-jffs2` via tftp at 192.168.1.10
* Interrupt uboot countdown, execute commands
  `run lk`
  `run lf`
  to flash the kernel / filesystem accordingly

MAC addresses as verified by OEM firmware:
use   address   source
LAN   *:fb      mib0 0x30 ('eth0mac'), art 0x1002 (label)
2g    *:fc      mib0 0x4b ('wifi0mac')
5g    *:fd      mib0 0x66 ('wifi1mac')

Signed-off-by: Sebastian Schaper <openwrt@sebastianschaper.net>

s-2 committed:
ath79: add support for ZyXEL NWA1123-AC

Specifications:
 * AR9342, 16 MiB Flash, 64 MiB RAM, 802.11n 2T2R, 2.4 GHz
 * QCA9882 PCIe card, 802.11ac 2T2R
 * 1x Gigabit Ethernet (AR8035), 802.3af PoE

Installation:
* OEM Web UI is at 192.168.1.2
  login as `admin` with password `1234`
* Flash factory-AAOX.bin

The string `AAOX` needs to be present within the file name of the uploaded
image to be accepted by the OEM Web-based updater, the factory image is
named accordingly to save the user from the hassle of manual renaming.

TFTP Recovery:
* Open the case, connect to TTL UART port (this is the official method
  described by Zyxel, the reset button is useless during power-on)
* Extract factory image (.tar.bz2), serve `vmlinux_mi124_f1e.lzma.uImage`
  and `mi124_f1e-jffs2` via tftp at 192.168.1.10
* Interrupt uboot countdown, execute commands
  `run lk`
  `run lf`
  to flash the kernel / filesystem accordingly

MAC addresses as verified by OEM firmware:
use   address   source
LAN   *:1c      mib0 0x30 ('eth0mac'), art 0x1002 (label)
2g    *:1c      mib0 0x4b ('wifi0mac')
5g    *:1e      mib0 0x66 ('wifi1mac')

Signed-off-by: Sebastian Schaper <openwrt@sebastianschaper.net>

s-2 commented:
Suggestions to improve the factory image generation are welcome, e.g. is there any more appropriate environment variabl (or a different way) to find the path for the loader image?

xabolcs commented:
You could look at the `senao-tar-gz` recipe for hints!

Btw, these devices really need the OKLI mode? How does the original partition look? The `uimage-lzma-loader` isn't enough?

s-2 commented:
> You could look at the `senao-tar-gz` recipe for hints!

Thanks, I'll have a look! It seems I had only checked `generic.mk` for similar devices...

> Btw, these devices really need the OKLI mode? How does the original partition look? The `uimage-lzma-loader` isn't enough?

The original partition layout looks like
```
...
0x000000050000-0x000000850000 : "rootfs"
0x000000850000-0x000000940000 : "uImage"
0x000000940000-0x000000fa0000 : "reserve"
...
```
with the `reserve` partition being unused (it is not blank, but nothing seems to be read / written there), but the `uImage` part is too small to hold the kernel. The OEM web updater can only flash the `rootfs` and `uImage` partitions, thus OKLI is needed (since rootfs comes before uImage) to support flashing OpenWrt without opening the case. On firstboot, the jffs2 partition will be expanded to use the full available flash space (`mtd-concat` over `rootfs` and the remainder of `uImage` + `reserve`).

So technically, this could still be extended to increase `IMAGE_SIZE` to include the remainder of `uImage`, i.e. cutoff the sysupgrade part > 8192k and concatenate to the loader image. I'll have a look at the `senao-tar-gz` first, thanks! :slightly_smiling_face:

xabolcs commented on target/linux/ath79/dts/ar9342_zyxel_nwa112x.dtsi:
```diff
@@ -0,0 +1,164 @@
+
+	virtual_flash {
+		compatible = "mtd-concat";
+		devices = <&firmware0 &firmware1>;
```
Please rename `firmware0` and `firmware1` to something which refers to "contatenation" too, e.g. `fwconcat0` and `fwconcat1`.

See commit 1fb40a72da115073b55ed1b92b443c1b80794c44 for the background!

xabolcs commented:
> The OEM web updater can only flash the `rootfs` and `uImage` partitions, thus ...

IS the OEM web updater able to write an ovesized (`uImage` + `reserve` sized) `vmlinux_mi124_f1e` to `uImage`?

s-2 commented:
> Is the OEM web updater able to write an ovesized (`uImage` + `reserve` sized) `vmlinux_mi124_f1e` to `uImage`?

No, the OEM updater is a shell script that calls `mtd write` for each file, which truncates the image at the end of the mtd part. (However, the recovery method would allow flashing oversized images, since uboot does not verify the partition boundaries).

So, when they user needs the full flash space for more packages in the squashfs, they could flash in two stages - a regular factory image followed by a customized sysupgrade, that can use all of the available 16 MiB flash (excl. boot / config stuff).


The environment variables `$(KDIR)/loader-$(DEVICE_NAME)` made it a lot easier, thanks! :+1: 

I'm just curious, is there a particular reason for having the `$(TAR)` command an environment variable? (e.g. differently named utilities on certain platforms etc.?)

Not sure whether creating a temporary directory for the archive contents is preferred over `--transform`'ing absolute to relative paths here.

xabolcs commented:
> No, the OEM updater is ...

OK. Just wanted to avoid the recipe boilerplate what the OKLI mode needs! :slightly_smiling_face: 
But a two stage install method is more boring than the OKLI boilerplate.

adschm commented on target/linux/ath79/image/generic.mk:
```diff
@@ -2229,6 +2237,31 @@ define Device/zbtlink_zbt-wd323
+
+define Device/zyxel_nwa1121-ni
+  $(Device/zyxel_nwa112x)
+  DEVICE_VENDOR := ZyXEL
```
DEVICE_VENDOR might be moved to the common node (as it's Zyxel-specific as it appears)

s-2 replied:
Agreed. The vendor is clearly visible from the included build recipe, thus vendor/model/variant are not scattered here.

mpratt14 commented:
> this could still be extended to increase `IMAGE_SIZE`

IMAGE_SIZE should be set to the size of `fwconcat0`, if the image is larger than that, booting will fail due to read-only rootfs being split between two partitions (if it doesn't get partially erased as well)


> OEM updater is a shell script

this is the same with Senao / Engenius products, specifically a file always with the same name `/etc/fwupgrade.sh`
would you mind posting it somewhere? like in a gist?


> is there a particular reason for having the `$(TAR)` command an environment variable?

for some makefiles, it specifies a different program to use, or specifically to use the compiled tar program in the toolchain instead of the host system's tar program, but here I think it includes some options


> `--transform`'ing absolute to relative paths

are you sure this option is needed? is it possible to just change directory instead?
including ` -C . ` in the command?

mpratt14 commented:
also, is there an ethernet PHY chip between transformer and SOC? like the AR8035 we have dealt with many time?

if so, that info should be included in the commit message Specification, along with flash and ram

s-2 commented:
> IMAGE_SIZE should be set to the size of `fwconcat0`, if the image is larger than that, booting will fail due to read-only rootfs being split between two partitions (if it doesn't get partially erased as well)

does this apply even for partitions aligned to a 64k boundary? At least the jffs2 has no issues, though that's only mounted after the "full" kernel has booted, not sure about limitations for the squashfs (however at least the kernel needs to be "in one piece" for OKLI to extract it). Currently it is limited to 8 MiB mostly to prevent an ugly mess of a factory image build recipe.

> this is the same with Senao / Engenius products, specifically a file always with the same name `/etc/fwupgrade.sh`
> would you mind posting it somewhere? like in a gist?

https://gist.github.com/s-2/b82db3b005391aab0656eb06ef5bc3b8
these scripts re located in /etc/scripts

> are you sure this option is needed? is it possible to just change directory instead?
> including `-C .` in the command?

I'll try what can be achieved by this, thanks. As far as I recall, the initial issue was that files added to the archive (from another directory) had the full pathname in the resulting archive, even when they were referenced by relative paths in the command arguments. So I could either create a new temporary directory and copy/move files there, then call tar on that directory and finally delete it, or truncate the paths while adding the files.

> also, is there an ethernet PHY chip between transformer and SOC?

There is an AR8031 or something, I'll add that.

s-2 commented:
This is weird, the paths are perfectly removed now even without the `--transform` or `-C .` options, it must have been related to the change towards using the `$(KDIR)` environment variable (or should I need even more than `make clean` to remove temporary build artifacts?)...

mpratt14 commented:
> does this apply even for partitions aligned to a 64k boundary? At least the jffs2 has no issues, though that's only mounted after the "full" kernel has booted, not sure about limitations for the squashfs

I believe it is a limitation of the mtd parser, but it may also be limitation of squashfs as well. I tried it once and it was unable to mount, but I don't remember whether it was identified correctly.



> AR8035

this is being picky now, but my experience with similar boards tells me there's a way to eliminate all the GMAC settings for delay, letting the PHY chip do all that work. you can experiment with different phy-mode and see what you can eliminate, and you can have wireless enabled by default in case of ethernet being dead, to maintain network access for the next flash.

starting with this

```
	pll-data = <0x02000000 0x00000101 0x00001313>;
```

and if that works then try without the TX delays in `gmac-config`

then if that breaks ethernet, change phy-mode to `"rgmii-rxid"`
but if it works, try without RX delay with phy-mode `"rgmii-txid"`
etc. etc.

also, this is something that usually varies between SOC, so not usually something in a DTSI

s-2 commented:
> pll-data = <0x02000000

tried setting to `0x02000000`, it kind of worked flawlessly, but the throughputs are not quite stable:

```
iperf3 -c 192.168.1.1
Connecting to host 192.168.1.1, port 5201
[  5] local 192.168.1.10 port 33374 connected to 192.168.1.1 port 5201
[ ID] Interval           Transfer     Bitrate         Retr  Cwnd
[  5]   0.00-1.00   sec  26.6 MBytes   223 Mbits/sec    0    221 KBytes       
[  5]   1.00-2.00   sec  26.4 MBytes   222 Mbits/sec    0    221 KBytes       
[  5]   2.00-3.00   sec  19.1 MBytes   161 Mbits/sec    0    221 KBytes       
[  5]   3.00-4.00   sec  16.4 MBytes   137 Mbits/sec    0    221 KBytes       
[  5]   4.00-5.00   sec  16.6 MBytes   139 Mbits/sec    0    221 KBytes       
[  5]   5.00-6.00   sec  16.1 MBytes   135 Mbits/sec    0    221 KBytes       
[  5]   6.00-7.00   sec  16.4 MBytes   138 Mbits/sec    0    221 KBytes       
[  5]   7.00-8.00   sec  17.2 MBytes   144 Mbits/sec    0    221 KBytes       
[  5]   8.00-9.00   sec  26.1 MBytes   219 Mbits/sec    0    221 KBytes       
[  5]   9.00-10.00  sec  26.8 MBytes   225 Mbits/sec    0    221 KBytes       
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-10.00  sec   208 MBytes   174 Mbits/sec    0             sender
[  5]   0.00-10.00  sec   207 MBytes   173 Mbits/sec                  receiver

```
When performing longer tests, the rate wiggles between ~220 and ~140 Mbit with intervals of roughly 10 seconds. Router being the server, PC is client (slightly slower the other way round, but also varying).

With `0x06000000` it looks way more stable:

```
iperf3 -c 192.168.1.1
Connecting to host 192.168.1.1, port 5201
[  5] local 192.168.1.10 port 34654 connected to 192.168.1.1 port 5201
[ ID] Interval           Transfer     Bitrate         Retr  Cwnd
[  5]   0.00-1.00   sec  26.8 MBytes   224 Mbits/sec    0    221 KBytes       
[  5]   1.00-2.00   sec  26.3 MBytes   220 Mbits/sec    0    221 KBytes       
[  5]   2.00-3.00   sec  26.3 MBytes   220 Mbits/sec    0    221 KBytes       
[  5]   3.00-4.00   sec  26.3 MBytes   221 Mbits/sec    0    221 KBytes       
[  5]   4.00-5.00   sec  26.6 MBytes   224 Mbits/sec    0    221 KBytes       
[  5]   5.00-6.00   sec  26.1 MBytes   219 Mbits/sec    0    221 KBytes       
[  5]   6.00-7.00   sec  26.5 MBytes   222 Mbits/sec    0    221 KBytes       
[  5]   7.00-8.00   sec  26.5 MBytes   222 Mbits/sec    0    221 KBytes       
[  5]   8.00-9.00   sec  26.4 MBytes   222 Mbits/sec    0    221 KBytes       
[  5]   9.00-10.00  sec  26.4 MBytes   221 Mbits/sec    0    221 KBytes       
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-10.00  sec   264 MBytes   222 Mbits/sec    0             sender
[  5]   0.00-10.00  sec   263 MBytes   221 Mbits/sec                  receiver
```

mpratt14 commented:
hmmm this is very interesting

the way I test stability is by doing things in SSH (dmesg, hexdump, etc.) and pinging large packets
maybe I should start using iperf 😆 

also, I wonder if `0x02000000` could be stable if phy-mode is set to `rgmii-rxid` (or `rgmii-txid`...not certain which)

however, it's possible that changing phy-mode incorrectly could break ethernet so if you try that be prepared with UART or wireless enabled by default...

s-2 commented:
setting `rgmii-rxid` with `0x02000000` (i.e. gmac tx delay disabled):
- no access at all

setting `rgmii-txid` with `0x02000000`:
- no access either

> the way I test stability is by doing things in SSH (dmesg, hexdump, etc.) and pinging large packets

Usually, when things were wrong about these settings, it was obvious already when just trying to access LuCI, i.e. everything was loading horribly slow etc., or nothing at all. In this case I have to admit, I wouldn't have seen any difference in the performance of LuCI etc. without using iperf, but then again this situation is usually rare.

By the way, I still have not figured out how to get gigabit working on DAP-2360 and DAP-2553 at all, maybe it's still an issue with ath79 drivers, not sure if [this comment](https://github.com/openwrt/openwrt/pull/1146#issuecomment-405065310) could be related.

The difference is that these devices are ar72xx, which does not have these gmac registers to work around this issue.

Maybe it was indeed working back in ar71xx, without touching ar93xx gmac delay settings (which could then probably make it work on ar72xx as well)? But I currently don't have the time and motivation to backport devices to ar71xx, only to verify this.

mpratt14 commented:
> The difference is that these devices are ar72xx, which does not have these gmac-config registers to work around this issue.

right...the ar724x SOC do not have GMAC configuration, but the PLL register to control the clock for RGMII is there
(at least for ar7242, not certain about the others)

if you can get to uboot console, ping with gigabit speed, and then do this
`md 0x1805002c 1`

then copy it to the first element of `pll-data` in DTS

it's possible that if you do it this way, `phy-mode` must be just `rgmii`
or you can look at the datasheet to rewrite the value without bits for delay

in total between DTS and DTSI there are 3 required lines
pll-data
pll-reg
pll-handle

and they are used in the driver here
https://github.com/openwrt/openwrt/blob/master/target/linux/ath79/files/drivers/net/ethernet/atheros/ag71xx/ag71xx_main.c#L1578

it may be that something in one of the DTSI is missing...

mpratt14 commented:
@s-2 i replied to your old forum post with many examples to try...

I had that list for a while, I should have posted it sooner 😆

s-2 commented:
> I had that list for a while, I should have posted it sooner laughing

actually, since I'm way too lazy I happen to have a [GUI application](http://sebastianschaper.net/openwrt/gmac_config.png) for this exact purpose, i.e. reading/writing registers from/to uboot via uart... scraping tooltips from datasheet bitfield tables... (hope this will be published one day :innocent: probably way more useful in the long term than messing with outdated routers in the first place)

s-2 commented:
(moved updated `tar` command to first commit)

CodeFetch commented on target/linux/ath79/dts/ar9342_zyxel_nwa112x.dtsi:
```diff
@@ -0,0 +1,165 @@
+/ {
+	aliases {
+		label-mac-device = &eth0;
+
```
Omit line break.

s-2 replied:
A significant majority of devices indeed don't have a newline here, so I'll omit this one. There seems to be no consensus however on whether label-mac-device should go to the top or bottom of aliases.

CodeFetch commented on target/linux/ath79/dts/ar9342_zyxel_nwa112x.dtsi:
```diff
@@ -0,0 +1,165 @@
+
+	virtual_flash {
+		compatible = "mtd-concat";
+
```
Omit line break.

s-2 replied:
While it seems a slight majority of ath79 devices have an empty line here, it's probably more consistent with general dts style to omit this one, though I don't have strong opinions on that (besides it is quite mixed for devices using `fwconcat`).

adschm replied:
I personally prefer to have compatible separated (as I do for status) and would put an empty-line here. But as for the others, this is purely taste.

CodeFetch reviewed (commented):
The rest looks good to me.

CodeFetch reviewed (approved):
Looks good.

adschm commented on target/linux/ath79/generic/base-files/etc/hotplug.d/ieee80211/10_fix_wifi_mac:
```diff
@@ -65,4 +65,7 @@ case "$board" in
 			mtd_get_mac_ascii u-boot-env ethaddr > /sys${DEVPATH}/macaddress
 		;;
+	zyxel,nwa1121-ni)
+		mtd_get_mac_text mib0 0x4b > /sys${DEVPATH}/macaddress
```
If you overwrite the WiFi MAC address here, we would you then use the original WiFi MAC address for ethernet?

s-2 replied:
`eth0mac` from `mib0` (at `0x30`) is indeed the same as the binary one located in `art` at `0x1002`, which is used for ethernet, while `wifi0mac` is located at `0x4b` in `mib0`.
Since there is only a dt binding for the binary format, only the ethernet mac can be set directly from the dts node (though it is indeed a little awkward that this is stored in the position that would usually hold wifi calibration data).

adschm replied:
Please preserve this information in a MAC address overview table in your commit message:
https://openwrt.org/docs/guide-developer/device-support-policies#mac_addresses

adschm commented on target/linux/ath79/image/generic.mk:
```diff
@@ -151,6 +152,12 @@ define Build/wrgg-pad-rootfs
 endef
 
+define Build/zyxel-tar-bz2
+	mv $@ $(word 2,$(1))
```
Please don't just use static file names (if I read this correctly), but use some $@.tmp folder where you can put them ...

s-2 replied:
While the exact file names to be put into the tar archive need to match what the updater is expecting, the tar operation itself may as well be executed on files in a temp folder (i.e. create a new folder, move files there, call tar and `rm -rd` the folder). I'll add that tomorrow.

mpratt14 replied:
good example here, if you need it

target/linux/ath79/image/common-senao.mk

adschm commented on target/linux/ath79/dts/ar9342_zyxel_nwa112x.dtsi:
```diff
@@ -0,0 +1,163 @@
+
+			partition@fe0000 {
+				label = "mib0";
+				reg = <0xfe0000 0x10000>;
```
Please use six digits consistently.

adschm commented on target/linux/ath79/image/generic.mk:
```diff
@@ -2229,6 +2236,31 @@ define Device/zbtlink_zbt-wd323
+  IMAGE_SIZE := 8192k
+  IMAGES += factory-$$$$(ZYXEL_MODEL_STRING).bin
+  IMAGE/factory-$$$$(ZYXEL_MODEL_STRING).bin := \
+	$$(IMAGE/sysupgrade.bin) | pad-to 8192k | zyxel-tar-bz2 \
```
Do not use sysupgrade.bin for factory. It will contain the metadata, which is unneeded here.

adschm commented on target/linux/ath79/generic/base-files/etc/hotplug.d/ieee80211/10_fix_wifi_mac:
```diff
@@ -65,7 +65,12 @@ case "$board" in
+			mtd_get_mac_text mib0 0x4b > /sys${DEVPATH}/macaddress
+		[ "$PHYNBR" -eq 1 ] && \
+			mtd_get_mac_text mib0 0x66 > /sys${DEVPATH}/macaddress
+		;;
```
Why this sudden change? And apart from that, this is broken due to the duplicate closing sign.

s-2 replied:
Is there any preferred way of incrementally adding cases for devices separated into multiple commits?
I will re-add `[ "$PHYNBR" -eq 0 ] && \` to the NWA1121, so that fewer changes are being introduced by the follow-up commit.
Is it generally acceptable in the first place to have a "redundant" `-eq` here (for NWA1121 that only has one wifi PHY), when it reduces the amount of case blocks?

adschm replied:
I think adding conditions that will never apply here will cause confusion. An additional device case does not really hurt, but will make things clear for everybody.

adv4ncr commented:
Would it be possible to also have support for the ZyXEL NAP102 AP ?

s-2 commented:
- separated case blocks for `10_fix_wifi_mac`
- consistent 0-padding for dts; re-added blank line after `compatible = mtd-concat`
- use `@.tmp` directory for tar archive generation
- omit metadata for factory image

s-2 commented:
> Would it be possible to also have support for the ZyXEL NAP102 AP ?

That device seems to use the same `zldfs` (ZyXEL dual image file system) format as the newer NWA series, eg. NWA 512x or 1123-A v2 / HD / ..., they are currently on my bucket list, right after finishing the COVR / EXO devices.

As soon as NWA5123 would be supported, NAP102 might be easy to add based on that.

CodeFetch commented:
@adv4ncr That could be a completely different hardware inside the case. Furthermore we've already found out that the managed versions (NWA5121 & NWA5123) of the NWA1121/NWA1123 have a different (more complex) firmware format while their hardware is identical. Sorry, but adding support for these ZyXEL devices is not easy and costs a lot of time. You may donate some of these devices to s-2 and he might give it a try I guess.

s-2 commented:
> You may donate some of these devices to s-2 and he might give it a try I guess.

It's not like a hardware donation would spare me any more time to work on that device though, also supporting this obviously assumes that the hardware will be supported, i.e. also based on Qualcomm or some other supported target - I had just checked the format of the firmware offered from the ZyXEL download page.

adschm commented:
And note that making the PR bigger will not really help to have it reviewed fast ...

adv4ncr commented:
> That device seems to use the same `zldfs` (ZyXEL dual image file system) format as the newer NWA series, eg. NWA 512x or 1123-AC v2 / HD / ..., they are currently on my bucket list, right after finishing the COVR / EXO devices.
> 
> As soon as NWA5123 would be supported, NAP102 might be easy to add based on that.

Really nice, thanks 😀 



> And note that making the PR bigger will not really help to have it reviewed fast ...

noted

adschm commented on target/linux/ath79/dts/ar9342_zyxel_nwa112x.dtsi:
```diff
@@ -0,0 +1,164 @@
+
+			fwconcat0: partition@50000 {
+				label = "fwconcat0";
+				reg = <0x50000 0x800000>;
```
0x050000

adschm commented on target/linux/ath79/image/generic.mk:
```diff
@@ -9,6 +9,7 @@ DEVICE_VARS += SEAMA_SIGNATURE SEAMA_MTDBLOCK
 DEVICE_VARS += KERNEL_INITRAMFS_PREFIX DAP_SIGNATURE
 DEVICE_VARS += EDIMAX_HEADER_MAGIC EDIMAX_HEADER_MODEL
 DEVICE_VARS += OPENMESH_CE_TYPE
+DEVICE_VARS += ZYXEL_MODEL_STRING
```
Consider adding this to the previous line ...

s-2 commented:
fixed formatting, included MAC address tables in commit messages.

Neustradamus commented:
To follow this PR.

adschm commented:
> To follow this PR.

Please don't spam.

vtmb commented:
When I try to compile it, i get this errormessage: luci.c:969:14 error: 'IWINFO_HTMODE_HE160' undeclared (first use in this function); did you mean 'IWINFO_HTMODE_VHT169'?

I checked out the commit regarding the NWA1123-AC

Neustradamus commented:
@s-2: Have you seen the @vtmb comment?

s-2 commented:
> luci.c:969:14 error: 'IWINFO_HTMODE_HE160' undeclared (first use in this function); did you mean 'IWINFO_HTMODE_VHT169'

This seems unrelated to the commit, maybe it needs to be rebased to master prior to building, assuming there might have been changes to the luci feed that require latest master to build.

vtmb commented:
@s-2 Are you still working on this or have you lost interest in this?

s-2 commented:
> or could you even provide compiled binary

I put some binaries here from the latest backup I could find (before some `make clean` wiped later results), these were built on April 20th, no significant changes to the image format were made after that:
[openwrt-ath79-generic-zyxel_nwa1123-ac-squashfs-factory-AAOX.bin](http://sebastianschaper.net/openwrt/openwrt-ath79-generic-zyxel_nwa1123-ac-squashfs-factory-AAOX.bin)
[openwrt-ath79-generic-zyxel_nwa1123-ac-squashfs-sysupgrade.bin](http://sebastianschaper.net/openwrt/openwrt-ath79-generic-zyxel_nwa1123-ac-squashfs-sysupgrade.bin)

To build a current image, just checkout latest OpenWrt master, add my branch as a remote (e.g. `git remote add github_s2 git@github.com:s-2/openwrt.git`) checkout the `github_s2/nwa112x_master` branch, then `git rebase master` before compiling.

vtmb commented:
Thank you! I will test the binary tomorrow and when i have some spare time i will also test compiling the binaries myself. I wasn't aware about these steps regarding rebase.

If this works, I can rescue at least 5 Accesspoints that have been shut down due to the old software running on them 👍🏻

CodeFetch commented:
@s-2 ART partition logic must be updated to nvmem approach:
https://github.com/openwrt/openwrt/commit/abc17bf306acd1c5954fbba97134e891439f917c

s-2 commented:
* updated `eth0` mac declaration to `nvmem-cells`

Cheddoleum commented:
I've installed your prebuilt binary on my recently-abandoned NWA1123-AC and am testing now.  

So far It's working faultlessly on every front: signal strength and coverage, performance and features.  Running POE with five SSIDs on both radios on three 802.1q VLANs. UI is far more responsive than stock.  Plus I finally get DFS channels. 

Very nice work; you've saved a much-better-than-average AP from the trash bin. Let me know if there's something you'd like tested.

Cheddoleum commented:
> I put some binaries here from the latest backup I could find (before some `make clean` wiped later results), these were built on April 20th, no significant changes to the image format were made after that:

Thanks for this. Curiously, when I build using defaults, only setting the target device, I get an image that is 1MB smaller than yours:  5604692 bytes for the factory .bin rather than yours' 6736998.  In addition, I see a very distinct difference in ath10k performance as noted above: a consistent ~210/~260mbps on my build vs ~260/~320 on yours under identical conditions; in both cases apparently CPU limited as sirq hits 99%.

Can you share your build config, or if it's not available, suggest anything that might be distinct about it compared to defaults? Both appear to be using the ct firmware; my build includes the ct-smallbuffers driver by default; I don't know what other factors might be important.

s-2 commented:
> Curiously, when I build using defaults, only setting the target device, I get an image that is 1MB smaller than yours

I usually include LuCI (and sometimes additional packages like iperf3 etc.), which is not part of the default build configuration.

While this would only make up for several hundred kilobytes, I also noticed that sometimes, after switching branches of different ages(?) etc. too often without deleting the whole config, I ended up with images lacking the required ath10k firmwares, even when these are declared in `DEVICE_PACKAGES`.
So I sometimes had to manually include these via menuconfig, thus I cannot guarantee that no unnecessary packages for other chips might still linger around in the firmware :) `opkg list-installed | grep ath10k` should show.

Apart from that, the only other information I could give is that the image was built back in April, so if the speed differences are indeed reproducible, maybe there is a slight regression involved with 2e10ed925e1e07c28570731a429efa5e7de3b826 ?
I'm currently on vacation and don't have a device with me for testing.

Cheddoleum commented:
> // edit: the speed regression indeed seems to be a side effect of FragAttack fixes: [greearb/ath10k-ct@b44cd7b](https://github.com/greearb/ath10k-ct/commit/b44cd7b2e7b0df5995ece18f358d4dfc40834ba1)

Yes, I see what you mean, thanks for spotting it. Oh well, security updates are the precisely reason we use OpenWRT rather than keep running EOL APs like this on their stock firmware. If the CPU impact can't be optimized out, it comes back to a choice between living with it or buying something more powerful that can cope.

Neustradamus commented:
What is the blocking part before a merging?

vtmb commented:
> What is the blocking part before a merging?

There needs to be approval. @adschm @mpratt14 @xabolcs 

Would really appreciate if this Accesspoint would be officially supported!

xabolcs commented:
@vtmb , sorry I'm not committer / reviewer just an OpenWrt fan :sweat_smile: ... can't help merging this.

vtmb commented:
> @vtmb , sorry I'm not committer / reviewer just an OpenWrt fan 😅 ... can't help merging this.

Oh sorry, I misunderstood the box listing you as reviewer. Nevermind!

adschm commented on target/linux/ath79/image/generic.mk:
```diff
@@ -2229,6 +2236,31 @@ define Device/zbtlink_zbt-wd323
+  IMAGE_SIZE := 8192k
+  IMAGES += factory-$$$$(ZYXEL_MODEL_STRING).bin
+  IMAGE/factory-$$$$(ZYXEL_MODEL_STRING).bin := \
+	$$(IMAGE/default) | pad-to 8192k | check-size | zyxel-tar-bz2 \
```
I don't think IMAGE/default is defined here?

s-2 replied:
Does this imply the value is currently taken from `Device/wd_mynet-n750`, which has it declared via `Device/seama`, thus (accidentally) creating a valid image?
i.e. this would break as soon as a new device is introduced in between the two, that overrides this with an incompatible definition? Interesting, I had initially considered this a boilerplate defined somewhere in the upper Makefile structures, will fix by the start of next week, thanks.

adschm replied:
I'm not quite sure. If per-device variables have no default value (e.g. in Device/Default), they indeed do take the last value set to them. The question is whether you tested factory flashing after you replaced the IMAGE/sysupgrade here, since the padding will produce an image with reasonable size even if IMAGE/default would just do nothing.

Anyway, just put the correct individual commands here ...

s-2 commented:
replaced use of `$$(IMAGE/default)` with explicit build steps declaration;
built, tested flashing both OEM -> OpenWrt and sysupgrade

s-2 commented:
rebased to master; switched to common recipe for OKLI (#4596)

s-2 commented:
rebased to master (because why not, it's been almost a year)
moved nvmem-cells mac address declaration directly into `art` partition node
tested on NWA1123-NI

DragonBluep commented:
The master branch has recently introduced some changes, see dae392752761128e13ede484217d806d643cbf2f 48625a04453b4f5775db2d2a5abebf977b81344c. And it looks like the calibration data for the PCIe NIC is missing?

s-2 commented:
> Now we can set ascii mac address in dts, please check if it works [ecd81de](https://github.com/openwrt/openwrt/commit/ecd81de7a5ab4899cc95107b402f7a402caf70bd) [5c147d3](https://github.com/openwrt/openwrt/commit/5c147d36ba7e998aee7ff4f81553b935bcfc5383)

Awesome, wasn't aware this is already possible by now, thanks! Will update & test tomorrow.

For 5GHz, there is indeed no caldata in the `art` partition; I'd assume it's due to the swappable miniPCIE card, using a dedicated eeprom (however MAC address is still set by the host).

vtmb commented:
It would be really great if you could provide binaries, I don't have a Linux box right now and apple gnu make is not compatible with your make files.

Happy to see progress on that, good that i still did not ditch my old access points!

s-2 commented:
Changed mac address assignment to nvmem-cells, but testing on NWA-1123-NI was not quite successful yet.
The 2.4 GHz mac address is correct, but for the 5 GHz card it is not, however it is not randomized on every boot, but constant (`00:30:AB:00:00:01`, maybe bogus MAC from card's built-in eeprom caldata?)

I also noticed that `qca,disable-5ghz` seems to be longer working since the rebase; did I miss anything there? Lots of devices in master still use it however. LuCI offers both bands for builtin radio, I'm quite sure this was different back then.
(actually, at some point I had both radios enabled, and the scan on the built-in 2.4G radio showed the 5 GHz AP with wrong MAC address)

DragonBluep commented:
@s-2 We can convert mtd-cal-data to nvmem-cells, see 48625a04453b4f5775db2d2a5abebf977b81344c.

> The 2.4 GHz mac address is correct, but for the 5 GHz card it is not, however it is not randomized on every boot, but constant (`00:30:AB:00:00:01`, maybe bogus MAC from card's built-in eeprom caldata?)

Maybe pcie id is not `pci168c,0030`? You can find this id in kernel log by searching `pci168c`. Or try to just remove `compatible = "pci168c,0030";`.

> I also noticed that `qca,disable-5ghz` seems to be longer working since the rebase; 

I saw someone use `ieee80211-freq-limit = <2400000 2500000>;` instead.

s-2 commented:
> I saw someone use `ieee80211-freq-limit = <2400000 2500000>;` instead.

Indeed, the `qca,disable-5ghz` property was removed 4c8dd973ef8e76f9fc622f0baab981769c89b601, though there are still two occurrences for one device (Araknis AN 300).

Now just the 5 GHz MAC address setting is not working, for both the `NWA1123-AC` and `-NI` devices.
The `qcom,ath10k` compatible should find the card, also the PCI VID/PID for the ath9k card matches `pci168c,0030`.

So I would assume setting the mac address this way only works when also caldata is provided via nvmem, not an internal eeprom? Maybe we still need `10_fix-wifi-mac` for these cases :slightly_frowning_face:

s-2 commented:
reverted the 5GHz mac address setting for now, to keep the PR in a non-broken state.

5GHz mac address is now correct again, 2.4G and Ethernet still working flawlessly via nvmem-cells.

s-2 committed:
ath79: add support for ZyXEL NWA1100-NH

Specifications:
 * AR9342, 16 MiB Flash, 64 MiB RAM, 802.11n 2T2R, 2.4 GHz
 * 1x Gigabit Ethernet (AR8035), 802.3af PoE

Installation:
* OEM Web UI is at 192.168.1.2
  login as `admin` with password `1234`
* Flash factory-AASI.bin

The string `AASI` needs to be present within the file name of the uploaded
image to be accepted by the OEM Web-based updater, the factory image is
named accordingly to save the user from the hassle of manual renaming.

TFTP Recovery:
* Open the case, connect to TTL UART port (this is the official method
  described by Zyxel, the reset button is useless during power-on)
* Extract factory image (.tar.bz2), serve `vmlinux_mi124_f1e.lzma.uImage`
  and `mi124_f1e-jffs2` via tftp at 192.168.1.10
* Interrupt uboot countdown, execute commands
  `run lk`
  `run lf`
  to flash the kernel / filesystem accordingly

MAC addresses as verified by OEM firmware:
use   address   source
LAN   *:cc      mib0 0x30 ('eth0mac'), art 0x1002 (label)
2g    *:cd      mib0 0x4b ('wifi0mac')

Signed-off-by: Sebastian Schaper <openwrt@sebastianschaper.net>

s-2 commented:
> It would be really great if you could provide binaries

These are built from the current state (2022-08-07), with 5 GHz set via legacy `10_fix_wifi_mac`.

I also wanted to add NWA1100-NH, which is almost identical to NWA1121-NI, the only difference being the amber status LED being red, and the green status LED being active high (so it's only working for the boot sequence, but off later; you can turn it on by setting it to "off" in LuCI :slightly_smiling_face: ).
I guess it would be way too much pain to refactor the .dtsi (add another cascade layer?) only for the LEDs...
However, this device has one additional info/wifi as well as another bi-color ethernet port LED, which is probably controlled by the AR8035 phy!? At least I could not identify the LED GPIOs yet.

[openwrt-ath79-generic-zyxel_nwa1121-ni-squashfs-factory-AABJ.bin](http://sebastianschaper.net/openwrt/openwrt-ath79-generic-zyxel_nwa1121-ni-squashfs-factory-AABJ.bin)
[openwrt-ath79-generic-zyxel_nwa1121-ni-squashfs-sysupgrade.bin](http://sebastianschaper.net/openwrt/openwrt-ath79-generic-zyxel_nwa1121-ni-squashfs-sysupgrade.bin)

[openwrt-ath79-generic-zyxel_nwa1123-ni-squashfs-factory-AAEO.bin](http://sebastianschaper.net/openwrt/openwrt-ath79-generic-zyxel_nwa1123-ni-squashfs-factory-AAEO.bin)
[openwrt-ath79-generic-zyxel_nwa1123-ni-squashfs-sysupgrade.bin](http://sebastianschaper.net/openwrt/openwrt-ath79-generic-zyxel_nwa1123-ni-squashfs-sysupgrade.bin)

[openwrt-ath79-generic-zyxel_nwa1123-ac-squashfs-factory-AAOX.bin](http://sebastianschaper.net/openwrt/openwrt-ath79-generic-zyxel_nwa1123-ac-squashfs-factory-AAOX.bin)
[openwrt-ath79-generic-zyxel_nwa1123-ac-squashfs-sysupgrade.bin](http://sebastianschaper.net/openwrt/openwrt-ath79-generic-zyxel_nwa1123-ac-squashfs-sysupgrade.bin)

[openwrt-ath79-generic-zyxel_nwa1100-nh-squashfs-factory-AASI.bin](http://sebastianschaper.net/openwrt/openwrt-ath79-generic-zyxel_nwa1100-nh-squashfs-factory-AASI.bin)
[openwrt-ath79-generic-zyxel_nwa1100-nh-squashfs-sysupgrade.bin](http://sebastianschaper.net/openwrt/openwrt-ath79-generic-zyxel_nwa1100-nh-squashfs-sysupgrade.bin)

mpratt14 commented:
AFAIK, `nvmem-cells` will only work if the MAC in flash is provided in binary, and `10_fix-wifi-mac` is for handling all other cases

for using the MAC of the internal eeprom, `nvmem-cells` would be omitted (but I don't know which chips that will or will not work for)

mpratt14 commented:
did you actually find `mac-address-ascii` in upstream?

s-2 commented:
> did you actually find `mac-address-ascii` in upstream?

It is implemented as a patch in OpenWrt, as suggested by https://github.com/openwrt/openwrt/pull/4037#issuecomment-1202988045; it works flawlessly for the 2.4G radio, but for 5GHz this device might be an exception, since a swappable miniPCIE card is used, thus containing a dedicated on-card EEPROM for caldata, which is (probably) why mac address setting does not work this way.

mpratt14 commented:
ah I see

maybe you can play with that patch by adding a bunch of printing to kernel log

and you're sure the address in `reg` is correct?

s-2 commented:
> maybe you can play with that patch by adding a bunch of printing to kernel log

I'd assume even with nvmem-cells it works the same way as before: the mac address would be patched into the caldata, before loading caldata into the chip. If there is no caldata being loaded from the host, there is no way to patch the MAC address this way.

> and you're sure the address in `reg` is correct?

At least it works the traditional way, using the same offset `0x66` from the `mib0` partition (which was confirmed to contain the 17 bytes 5GHz ASCII mac).

DragonBluep commented:
> If there is no caldata being loaded from the host, there is no way to patch the MAC address this way.

I think so too. It doesn't seem to always work correctly. By the way, `mac-address-ascii` doesn't work for mt7621's built-in switch. #9780

> I also wanted to add NWA1100-NH, which is almost identical to NWA1121-NI, the only difference being the amber status LED being red, and the green status LED being active high (so it's only working for the boot sequence, but off later; you can turn it on by setting it to "off" in LuCI 🙂 ).
> I guess it would be way too much pain to refactor the .dtsi (add another cascade layer?) only for the LEDs...

We can override these values in the `NWA1100-NH.dts`, but I'm not sure if this is the right way. At least we always override pcie status with `status = "okay";`.

s-2 commented:
- renamed .dtsi to `ar9342_zyxel_nwa11xx.dtsi`
- moved `leds` node and red/amber led `aliases` to individual .dts files
- added NWA1100-NH

I wonder whether it's acceptable to split the `aliases` node, or should I copy it redudantly to all individual `.dts` files?

Neustradamus commented:
@s-2: Excellent!
Can you edit the PR title too?

CodeFetch commented:
Looks good to me... @s-2 The only issue that remains is setting the MAC address of the external card, right?
@mpratt14 Can you review this once more?

mpratt14 commented:
> > and you're sure the address in `reg` is correct?

> At least it works the traditional way, using the same offset `0x66` from the `mib0` partition


sorry, I was referring to the PCIe address of the ath10k chip, which (I think) must be correct in the value of `reg` for the node within `&pcie` in order for any other binding value within that node to work (like nvmem-cell)

usually it is all 0's but its good to double check

the entire address should be visible from kernel log and by the pseudofs

examples:


```
$ dmesg | grep 168c
[    1.224992] pci 0000:00:00.0: [168c:003c] type 00 class 0x028000
```

```
$ ls /sys/bus/pci/drivers/ath10k_pci
0000:00:00.0  bind          module        new_id        remove_id     uevent        unbind
```

and if that already checks out then I guess we are stuck with the wifi fixup script...




> I wonder whether it's acceptable to split the `aliases` node

no, when I tried this Adrian ended up redoing my whole commit 😆

s-2 commented:
```[    1.154045] PCI host bridge to bus 0000:00
[    1.158265] pci_bus 0000:00: root bus resource [mem 0x10000000-0x13ffffff]
[    1.165243] pci_bus 0000:00: root bus resource [io  0x0000]
[    1.170921] pci_bus 0000:00: root bus resource [??? 0x00000000 flags 0x0]
[    1.177825] pci_bus 0000:00: No busn resource found for root bus, will use [bus 00-ff]
[    1.185942] pci 0000:00:00.0: [168c:003c] type 00 class 0x028000
[    1.192073] pci 0000:00:00.0: reg 0x10: [mem 0x00000000-0x001fffff 64bit]
[    1.199034] pci 0000:00:00.0: reg 0x30: [mem 0x00000000-0x0000ffff pref]
[    1.205935] pci 0000:00:00.0: supports D1 D2
[    1.210271] pci 0000:00:00.0: PME# supported from D0 D1 D2 D3hot D3cold
```
```
root@OpenWrt:~# ls /sys/bus/pci/drivers/ath10k_pci
0000:00:00.0  bind          module        new_id        remove_id     uevent        unbind
```


> no, when I tried this Adrian ended up redoing my whole commit :laughing:

:+1: will fix, thanks!

s-2 commented:
- moved `aliases` nodes to individual .dts
- renamed `nwa112x` buildstep include to `nwa11xx`

hauke commented:
Thank you for your patches, I applied them to master in 251ecfe37914919422d74dd32fab15a86f65e879, 527be5a456cecdc4e3296816906e117ec04bfdff, a6e0ca96da45067026f9c74ed1e0e16fc1cf7b7c and a434795809bb0759947f57117a28cd0df5e19e76.
